src/module.c

```diff
@@ -7389,6 +7415,184 @@ RedisModuleBlockedClient *moduleBlockClient(RedisModuleCtx *ctx, RedisModuleCmdF
+            if (result == REDISMODULE_AUTH_HANDLED) break;
+        }
+        if (cur_auth_ctx == c->custom_auth_ctx) {
+            skipped_prev_callbacks = 1;
+        }
+    }
+    return result;
+}
+
+/* Helper function to handle a reprocessed unblocked client that had custom auth ongoing.
+ * Returns REDISMODULE_AUTH_NOT_HANDLED if the client was not reprocessed after a blocking custom
+ * auth operation.
+ * Otherwise, we attempt the auth reply callback & the free priv data callback, update fields and
+ * return the result of the reply callback. */
+int attemptBlockedAuthReplyCallback(client *c, robj *username, robj *password, const char **err) {
+    int result = REDISMODULE_AUTH_NOT_HANDLED;
+    if (!c->module_blocked_client) return result;
+    RedisModuleBlockedClient *bc = (RedisModuleBlockedClient *) c->module_blocked_client;
+    bc->client = c;
+    if (bc->auth_reply_cb) {
+        RedisModuleCtx ctx;
+        moduleCreateContext(&ctx, bc->module, REDISMODULE_CTX_BLOCKED_REPLY);
+        ctx.blocked_privdata = bc->privdata;
+        ctx.blocked_ready_key = NULL;
+        ctx.client = bc->client;
+        ctx.blocked_client = bc;
+        result = bc->auth_reply_cb(&ctx, username, password, err);
+    }
+    moduleInvokeFreePrivDataCallback(c, bc);
+    c->module_blocked_client = NULL;
+    c->lastcmd->microseconds += bc->background_duration;
+    bc->module->blocked_clients--;
+    zfree(bc);
+    return result;
+}
+
+/* Helper function to attempt Module based authentication through custom auth callbacks.
+ * Here, the Module is expected to authenticate the client using the RedisModule APIs and to add ACL
+ * logs incase of errors.
+ * Returns C_ERR if no callbacks have been registered OR if a callback denied authentication OR if
+ * none of the callbacks successfully authenticated the client. In these cases, an ERR can be
+ * returned to the client.
+ * Returns C_OK if:
+ * - A callback has authenticated the client (concluding module based auth). Here, an OK message can
+ * be returned to the client.
+ * - Custom auth is still in progress through a blocking implementation. In this case, the client
+ * can be replied to later after the blocking operation & after handling the module client unblock.
+ */
+int checkModuleAuthentication(client *c, robj *username, robj *password, const char **err) {
+    if (!listLength(moduleCustomAuthCallbacks)) return C_ERR;
+    int result = attemptBlockedAuthReplyCallback(c, username, password, err);
+    if (result == REDISMODULE_AUTH_HANDLED) {
+        goto handle_result;
+    }
+    result = attemptNextCustomAuthCb(c, username, password, err);
+handle_result:
+    if (c->flags & CLIENT_BLOCKED) {
+        /* Modules are expected to return REDISMODULE_AUTH_HANDLED when blocking clients. */
+        serverAssert(result == REDISMODULE_AUTH_HANDLED);
+        return C_OK;
+    }
+    if (c->flags & CLIENT_PENDING_COMMAND) {
+        c->flags &= ~CLIENT_PENDING_COMMAND;
+    }
```

Comment:
From what I see this is needed for a proper clean up.

The `CLIENT_PENDING_COMMAND ` flag is added in `RM_BlockClientOnAuth`, but there are cases where blocking does not go through successfully. 
The one case that I know of is if the API was used in the middle of MULTI/LUA. It is needed for any other case like this where the client does not actually go through the normal blocking / unblocking flow.

It can be avoided by only adding the flag in `RM_BlockClientOnAuth` if the client was successfully blocked.
```
RedisModuleBlockedClient *RM_BlockClientOnAuth(RedisModuleCtx *ctx, RedisModuleCustomAuthCallback reply_callback,
                                               void (*free_privdata)(RedisModuleCtx*,void*)) {
    if (!clientHasModuleAuthInProgress(ctx->client)) {
        addReplyError(ctx->client, "Module blocking client on auth when not currently undergoing module authentication");
        return NULL;
    }
    RedisModuleBlockedClient *bc = moduleBlockClient(ctx,NULL,reply_callback,NULL,free_privdata,0, NULL,0,NULL,0);
    if (ctx->client->flags & CLIENT_BLOCKED) {
        ctx->client->flags |= CLIENT_PENDING_COMMAND;
    }
    return bc;
}
```

Reply:
I made this change here: https://github.com/redis/redis/pull/11659/commits/b7d12ca4ab0138c69c80d6b59e16ab2e66902816